pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+var (
+	errNilMembersCache = errors.New("member cache is nil")
+	errKeyExist        = errors.New("key already exist")
+	errKeyNotExist     = errors.New("key does not exist")
+)
+
+type membersCache struct {
+	mux   sync.RWMutex
+	cache map[string]etcdserverpb.Member
+}
+
+func newMembersCache() *membersCache {
+	return &membersCache{
+		mux:   sync.RWMutex{},
+		cache: make(map[string]etcdserverpb.Member),
+	}
+}
+
+// EtcdMemberListerController keeps etcd members cached, and updates the cache when the membership changes
+type EtcdMemberListerController struct {
+	membersCache   *membersCache
+	operatorClient v1helpers.OperatorClient
+	etcdClient     etcdcli.EtcdClient
+}
+
+func NewEtcdMemberListerController(operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &EtcdMemberListerController{
+		operatorClient: operatorClient,
+		etcdClient:     etcdClient,
+		membersCache:   newMembersCache(),
+	}
+	return factory.New().ResyncEvery(30*time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller"))
+}
+
+func (c *EtcdMemberListerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncEtcdMembers(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdMemberListerControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorSyncingEtcdMembers",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "EtcdMemberListerControllerDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "EtcdMembersSynced",
+	}))
+	if updateErr != nil {
+		syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *EtcdMemberListerController) syncEtcdMembers(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+
+	var syncErrs []error
+	// build map from current live members
+	currentMembersMap, errs := c.currentMembersMap(members)
+	if len(errs) > 0 {
+		syncErrs = append(syncErrs, errs...)
+	}
+
+	// check membership changes against members cache
+	if !reflect.DeepEqual(currentMembersMap, c.membersCache) {
+		klog.V(2).Infof("detected changes in etcd cluster membership, updating controller cache")
```

Comment:
can you add the previous members and the current members? I've done something similar in the client and it's super useful when going through must gathers